From 915f298fe217028f241b826ffdff347e391616dc Mon Sep 17 00:00:00 2001
From: Yu Qiaowei <cerf.yu@rock-chips.com>
Date: Fri, 10 Oct 2025 19:14:54 +0800
Subject: [PATCH 30/39] video: rockchip: rga3: enable config_intr and support
 parse error status

Change-Id: I00030dd532f417914183b1654fdc0b1aad8cd271
Signed-off-by: Yu Qiaowei <cerf.yu@rock-chips.com>
(cherry picked from commit ae8e8c4da037db91fb8c50e70a5d1cb5d63300af)
---
 .../rockchip/rga3/include/rga2_reg_info.h     | 20 ++++++++++++++++---
 drivers/video/rockchip/rga3/include/rga_drv.h |  1 +
 drivers/video/rockchip/rga3/rga2_reg_info.c   | 12 +++++++++++
 3 files changed, 30 insertions(+), 3 deletions(-)

diff --git a/drivers/video/rockchip/rga3/include/rga2_reg_info.h b/drivers/video/rockchip/rga3/include/rga2_reg_info.h
index f19a2421b888..2500a6996a65 100644
--- a/drivers/video/rockchip/rga3/include/rga2_reg_info.h
+++ b/drivers/video/rockchip/rga3/include/rga2_reg_info.h
@@ -20,6 +20,7 @@
 #define RGA2_MMU_CMD_BASE			0x018
 #define RGA2_STATUS2				0x01c
 #define RGA2_WORK_CNT				0x020
+#define RGA2_INTR_STATUS2			0x024
 #define RGA2_VERSION_NUM			0x028
 #define RGA2_READ_LINE_CNT			0x030
 #define RGA2_WRITE_LINE_CNT			0x034
@@ -153,6 +154,9 @@
 #define v_RGA2_STATUS1_SW_CMD_CUR_NUM(x)		(((x) >> 8) & 0xfff)
 
 /*RGA_INT*/
+#define m_RGA2_INT_CONFIG_ERR_CLEAR			(1 << 27)
+#define m_RGA2_INT_CONFIG_ERR_EN			(1 << 26)
+#define m_RGA2_INT_CONFIG_ERR				(1 << 25)
 #define m_RGA2_INT_FBCIN_DEC_ERROR_CLEAR		(1 << 24)
 #define m_RGA2_INT_FBCIN_DEC_ERROR_EN			(1 << 23)
 #define m_RGA2_INT_FBCIN_DEC_ERROR			(1 << 22)
@@ -184,23 +188,27 @@
 		m_RGA2_INT_MMU_INT_FLAG | \
 		m_RGA2_INT_ERROR_INT_FLAG | \
 		m_RGA2_INT_SCL_ERROR_INTR | \
-		m_RGA2_INT_FBCIN_DEC_ERROR \
+		m_RGA2_INT_FBCIN_DEC_ERROR | \
+		m_RGA2_INT_CONFIG_ERR \
 	)
 #define m_RGA2_INT_ERROR_CLEAR_MASK \
 	( \
 		m_RGA2_INT_MMU_INT_CLEAR | \
 		m_RGA2_INT_ERROR_INT_CLEAR | \
 		m_RGA2_INT_SCL_ERROR_CLEAR | \
-		m_RGA2_INT_FBCIN_DEC_ERROR_CLEAR \
+		m_RGA2_INT_FBCIN_DEC_ERROR_CLEAR | \
+		m_RGA2_INT_CONFIG_ERR_CLEAR \
 )
 #define m_RGA2_INT_ERROR_ENABLE_MASK \
 	( \
 		m_RGA2_INT_MMU_INT_EN | \
 		m_RGA2_INT_ERROR_INT_EN | \
 		m_RGA2_INT_SCL_ERROR_EN | \
-		m_RGA2_INT_FBCIN_DEC_ERROR_EN \
+		m_RGA2_INT_FBCIN_DEC_ERROR_EN | \
+		m_RGA2_INT_CONFIG_ERR_EN \
 	)
 
+#define s_RGA2_INT_CONFIG_ERR_CLEAR(x)			((x & 0x1) << 27)
 #define s_RGA2_INT_LINE_WR_CLEAR(x)			((x & 0x1) << 16)
 #define s_RGA2_INT_LINE_RD_CLEAR(x)			((x & 0x1) << 15)
 #define s_RGA2_INT_LINE_WR_EN(x)			((x & 0x1) << 14)
@@ -220,6 +228,12 @@
 #define m_RGA2_STATUS2_BUS_ERROR			(0x1 << 1)
 #define m_RGA2_STATUS2_RPP_ERROR			(0x1 << 0)
 
+/* RGA2_INTR_STATUS2 */
+#define m_RGA2_INTR_STATUS2_SRC_DST_RECT_NOT_EQUAL	(0x1 << 0)
+#define m_RGA2_INTR_STATUS2_SRC1_HORI_BND_ERR		(0x1 << 1)
+#define m_RGA2_INTR_STATUS2_SRC1_VERT_BND_ERR		(0x1 << 2)
+#define m_RGA2_INTR_STATUS2_SRC1_ODD_VIOLATION		(0x1 << 3)
+
 /* RGA_READ_LINE_CNT_TH */
 #define m_RGA2_READ_LINE_SW_INTR_LINE_RD_TH		(0x1fff << 0)
 
diff --git a/drivers/video/rockchip/rga3/include/rga_drv.h b/drivers/video/rockchip/rga3/include/rga_drv.h
index d3a6c93f2292..45c7afef77f4 100644
--- a/drivers/video/rockchip/rga3/include/rga_drv.h
+++ b/drivers/video/rockchip/rga3/include/rga_drv.h
@@ -338,6 +338,7 @@ struct rga_job {
 	uint32_t intr_status;
 	uint32_t hw_status;
 	uint32_t cmd_status;
+	uint32_t intr_status2;
 
 	uint32_t work_cycle;
 };
diff --git a/drivers/video/rockchip/rga3/rga2_reg_info.c b/drivers/video/rockchip/rga3/rga2_reg_info.c
index 24cdae11628a..321492c1b572 100644
--- a/drivers/video/rockchip/rga3/rga2_reg_info.c
+++ b/drivers/video/rockchip/rga3/rga2_reg_info.c
@@ -3373,6 +3373,7 @@ static int rga2_read_status(struct rga_job *job, struct rga_scheduler_t *schedul
 	job->hw_status = rga_read(RGA2_STATUS2, scheduler);
 	job->cmd_status = rga_read(RGA2_STATUS1, scheduler);
 	job->work_cycle = rga_read(RGA2_WORK_CNT, scheduler);
+	job->intr_status2 = rga_read(RGA2_INTR_STATUS2, scheduler);
 
 	return 0;
 }
@@ -3506,6 +3507,17 @@ static int rga2_isr_thread(struct rga_job *job, struct rga_scheduler_t *schedule
 		} else if (job->intr_status & m_RGA2_INT_FBCIN_DEC_ERROR) {
 			rga_job_err(job, "FBC decode failed, please check if the source data is FBC data.\n");
 			job->ret = -EACCES;
+		} else if (job->intr_status & m_RGA2_INT_CONFIG_ERR) {
+			rga_job_err(job, "reg config error. status[%#x]\n", job->intr_status2);
+			if (job->intr_status2 & m_RGA2_INTR_STATUS2_SRC_DST_RECT_NOT_EQUAL)
+				rga_job_err(job, "reg config error: src_rect != dst_rect.\n");
+			if (job->intr_status2 & m_RGA2_INTR_STATUS2_SRC1_HORI_BND_ERR)
+				rga_job_err(job, "reg config error: src1 horizontal beyond the boundary in overlay.\n");
+			if (job->intr_status2 & m_RGA2_INTR_STATUS2_SRC1_VERT_BND_ERR)
+				rga_job_err(job, "reg config error: src1 vertical beyond the boundary in overlay.\n");
+			if (job->intr_status2 & m_RGA2_INTR_STATUS2_SRC1_ODD_VIOLATION)
+				rga_job_err(job, "reg config error: src1 odd violation in overlay.\n");
+			job->ret = -EACCES;
 		}
 
 		if (job->ret == 0) {
